encoders/src/main/scala/org/apache/spark/sql/execution/columnar/ColumnBatch.scala

```diff
@@ -0,0 +1,248 @@
+    val delta2 = buffers.get(deltaPosition - 1)
+    if ((delta1 ne null) || (delta2 ne null)) {
+      UpdatedColumnDecoder(decoder, field, delta1, delta2)
+    } else null
+  }
+
+  def getDeletedColumnDecoder: ColumnDeleteDecoder = {
+    colBuffers.get(ColumnFormatEntry.DELETE_MASK_COL_INDEX) match {
+      case null => null
+      case deleteBuffer => new ColumnDeleteDecoder(deleteBuffer)
+    }
+  }
+
+  def getDeletedRowCount: Int = {
+    val delete = colBuffers.get(ColumnFormatEntry.DELETE_MASK_COL_INDEX)
+    if (delete eq null) 0
+    else {
+      val allocator = ColumnEncoding.getAllocator(delete)
+      ColumnEncoding.readInt(allocator.baseObject(delete),
+        allocator.baseOffset(delete) + delete.position() + 8)
+    }
+  }
+
+  private def releaseColumns(): Unit = {
+    val buffers = colBuffers
+    // not null check in case constructor itself fails due to low memory
+    if ((buffers ne null) && buffers.size() > 0) {
+      buffers.forEachValue(new Procedure[ByteBuffer] {
+        override def value(buffer: ByteBuffer): Unit = {
+          // release previous set of buffers immediately
+          if (buffer ne null) {
+            // release from accounting if decompressed buffer
+            if (!BufferAllocator.releaseBuffer(buffer) &&
+                (buffer.order() eq ByteOrder.LITTLE_ENDIAN)) {
+              releaseExecutionMemory(buffer, CompressionUtils.DECOMPRESSION_OWNER)
+            }
+          }
+        }
+      })
+      colBuffers = null
+    }
+  }
+
+  private def readColumnData(): Unit = {
+    val columnIndex = rs.getInt(3)
+    val columnBlob = rs.getBlob(4)
+    val columnBuffer = getBufferFromBlob(columnBlob)
+    if (columnBuffer ne null) {
+      // put all the read buffers in "colBuffers" to free on next() or close()
+      colBuffers.put(columnIndex, columnBuffer)
+      columnIndex match {
+        case ColumnFormatEntry.STATROW_COL_INDEX => currentStats = columnBuffer
+        case ColumnFormatEntry.DELTA_STATROW_COL_INDEX => currentDeltaStats = columnBuffer
+        case _ =>
+      }
+    }
+  }
+
+  override protected def moveNext(): Boolean = {
+    currentStats = null
+    currentDeltaStats = null
+    releaseColumns()
+    if (rsHasNext) {
+      currentUUID = rs.getLong(1)
+      // create a new map instead of clearing old one to help young gen GC
+      colBuffers = new IntObjectHashMap[ByteBuffer](totalColumns + 1)
+      // keep reading next till its still part of current column batch; if UUID changes
+      // then next call to "moveNext" will read from incremented cursor position
+      // else all rows may have been read which is indicated by "rsHasNext"
+      do {
+        readColumnData()
+        rsHasNext = rs.next()
+      } while (rsHasNext && rs.getLong(1) == currentUUID)
+      true
+    } else false
+  }
+
+  override protected def getCurrentValue: ByteBuffer = currentStats
+
+  override def close(): Unit = {
+    releaseColumns()
+    super.close()
+  }
+}
```

Comment:
confirm that this code is identical to current master

Reply:
Cross verified for being identical to the latest master